// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: center/user.proto

package common

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on RegisterReq with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *RegisterReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RegisterReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in RegisterReqMultiError, or
// nil if none found.
func (m *RegisterReq) ValidateAll() error {
	return m.validate(true)
}

func (m *RegisterReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetUsername()); l < 4 || l > 20 {
		err := RegisterReqValidationError{
			field:  "Username",
			reason: "value length must be between 4 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if l := utf8.RuneCountInString(m.GetPassword()); l < 6 || l > 20 {
		err := RegisterReqValidationError{
			field:  "Password",
			reason: "value length must be between 6 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetPhone() <= 0 {
		err := RegisterReqValidationError{
			field:  "Phone",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return RegisterReqMultiError(errors)
	}
	return nil
}

// RegisterReqMultiError is an error wrapping multiple validation errors
// returned by RegisterReq.ValidateAll() if the designated constraints aren't met.
type RegisterReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RegisterReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RegisterReqMultiError) AllErrors() []error { return m }

// RegisterReqValidationError is the validation error returned by
// RegisterReq.Validate if the designated constraints aren't met.
type RegisterReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RegisterReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RegisterReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RegisterReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RegisterReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RegisterReqValidationError) ErrorName() string { return "RegisterReqValidationError" }

// Error satisfies the builtin error interface
func (e RegisterReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRegisterReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RegisterReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RegisterReqValidationError{}

// Validate checks the field values on UsernameReq with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UsernameReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UsernameReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in UsernameReqMultiError, or
// nil if none found.
func (m *UsernameReq) ValidateAll() error {
	return m.validate(true)
}

func (m *UsernameReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetUsername()); l < 4 || l > 20 {
		err := UsernameReqValidationError{
			field:  "Username",
			reason: "value length must be between 4 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if l := utf8.RuneCountInString(m.GetPassword()); l < 6 || l > 20 {
		err := UsernameReqValidationError{
			field:  "Password",
			reason: "value length must be between 6 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UsernameReqMultiError(errors)
	}
	return nil
}

// UsernameReqMultiError is an error wrapping multiple validation errors
// returned by UsernameReq.ValidateAll() if the designated constraints aren't met.
type UsernameReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UsernameReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UsernameReqMultiError) AllErrors() []error { return m }

// UsernameReqValidationError is the validation error returned by
// UsernameReq.Validate if the designated constraints aren't met.
type UsernameReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UsernameReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UsernameReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UsernameReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UsernameReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UsernameReqValidationError) ErrorName() string { return "UsernameReqValidationError" }

// Error satisfies the builtin error interface
func (e UsernameReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUsernameReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UsernameReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UsernameReqValidationError{}

// Validate checks the field values on PhoneReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PhoneReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PhoneReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PhoneReqMultiError, or nil
// if none found.
func (m *PhoneReq) ValidateAll() error {
	return m.validate(true)
}

func (m *PhoneReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetPhone() <= 0 {
		err := PhoneReqValidationError{
			field:  "Phone",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return PhoneReqMultiError(errors)
	}
	return nil
}

// PhoneReqMultiError is an error wrapping multiple validation errors returned
// by PhoneReq.ValidateAll() if the designated constraints aren't met.
type PhoneReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PhoneReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PhoneReqMultiError) AllErrors() []error { return m }

// PhoneReqValidationError is the validation error returned by
// PhoneReq.Validate if the designated constraints aren't met.
type PhoneReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PhoneReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PhoneReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PhoneReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PhoneReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PhoneReqValidationError) ErrorName() string { return "PhoneReqValidationError" }

// Error satisfies the builtin error interface
func (e PhoneReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPhoneReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PhoneReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PhoneReqValidationError{}

// Validate checks the field values on EditReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *EditReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on EditReq with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in EditReqMultiError, or nil if none found.
func (m *EditReq) ValidateAll() error {
	return m.validate(true)
}

func (m *EditReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := EditReqValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(m.GetContent()) < 3 {
		err := EditReqValidationError{
			field:  "Content",
			reason: "value length must be at least 3 bytes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return EditReqMultiError(errors)
	}
	return nil
}

// EditReqMultiError is an error wrapping multiple validation errors returned
// by EditReq.ValidateAll() if the designated constraints aren't met.
type EditReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m EditReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m EditReqMultiError) AllErrors() []error { return m }

// EditReqValidationError is the validation error returned by EditReq.Validate
// if the designated constraints aren't met.
type EditReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EditReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EditReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EditReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EditReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EditReqValidationError) ErrorName() string { return "EditReqValidationError" }

// Error satisfies the builtin error interface
func (e EditReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEditReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EditReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EditReqValidationError{}

// Validate checks the field values on EditPwdReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *EditPwdReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on EditPwdReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in EditPwdReqMultiError, or
// nil if none found.
func (m *EditPwdReq) ValidateAll() error {
	return m.validate(true)
}

func (m *EditPwdReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := EditPwdReqValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if l := utf8.RuneCountInString(m.GetOldPwd()); l < 6 || l > 20 {
		err := EditPwdReqValidationError{
			field:  "OldPwd",
			reason: "value length must be between 6 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if l := utf8.RuneCountInString(m.GetPwd()); l < 6 || l > 20 {
		err := EditPwdReqValidationError{
			field:  "Pwd",
			reason: "value length must be between 6 and 20 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return EditPwdReqMultiError(errors)
	}
	return nil
}

// EditPwdReqMultiError is an error wrapping multiple validation errors
// returned by EditPwdReq.ValidateAll() if the designated constraints aren't met.
type EditPwdReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m EditPwdReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m EditPwdReqMultiError) AllErrors() []error { return m }

// EditPwdReqValidationError is the validation error returned by
// EditPwdReq.Validate if the designated constraints aren't met.
type EditPwdReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e EditPwdReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e EditPwdReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e EditPwdReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e EditPwdReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e EditPwdReqValidationError) ErrorName() string { return "EditPwdReqValidationError" }

// Error satisfies the builtin error interface
func (e EditPwdReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sEditPwdReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = EditPwdReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = EditPwdReqValidationError{}

// Validate checks the field values on UIDReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UIDReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UIDReq with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in UIDReqMultiError, or nil if none found.
func (m *UIDReq) ValidateAll() error {
	return m.validate(true)
}

func (m *UIDReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetId() <= 0 {
		err := UIDReqValidationError{
			field:  "Id",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UIDReqMultiError(errors)
	}
	return nil
}

// UIDReqMultiError is an error wrapping multiple validation errors returned by
// UIDReq.ValidateAll() if the designated constraints aren't met.
type UIDReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UIDReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UIDReqMultiError) AllErrors() []error { return m }

// UIDReqValidationError is the validation error returned by UIDReq.Validate if
// the designated constraints aren't met.
type UIDReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UIDReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UIDReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UIDReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UIDReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UIDReqValidationError) ErrorName() string { return "UIDReqValidationError" }

// Error satisfies the builtin error interface
func (e UIDReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUIDReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UIDReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UIDReqValidationError{}

// Validate checks the field values on UIDsReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UIDsReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UIDsReq with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in UIDsReqMultiError, or nil if none found.
func (m *UIDsReq) ValidateAll() error {
	return m.validate(true)
}

func (m *UIDsReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetIds()) < 1 {
		err := UIDsReqValidationError{
			field:  "Ids",
			reason: "value must contain at least 1 item(s)",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return UIDsReqMultiError(errors)
	}
	return nil
}

// UIDsReqMultiError is an error wrapping multiple validation errors returned
// by UIDsReq.ValidateAll() if the designated constraints aren't met.
type UIDsReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UIDsReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UIDsReqMultiError) AllErrors() []error { return m }

// UIDsReqValidationError is the validation error returned by UIDsReq.Validate
// if the designated constraints aren't met.
type UIDsReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UIDsReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UIDsReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UIDsReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UIDsReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UIDsReqValidationError) ErrorName() string { return "UIDsReqValidationError" }

// Error satisfies the builtin error interface
func (e UIDsReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUIDsReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UIDsReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UIDsReqValidationError{}

// Validate checks the field values on OnlineReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OnlineReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OnlineReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OnlineReqMultiError, or nil
// if none found.
func (m *OnlineReq) ValidateAll() error {
	return m.validate(true)
}

func (m *OnlineReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetServer()) < 3 {
		err := OnlineReqValidationError{
			field:  "Server",
			reason: "value length must be at least 3 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if utf8.RuneCountInString(m.GetToken()) < 3 {
		err := OnlineReqValidationError{
			field:  "Token",
			reason: "value length must be at least 3 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if len(errors) > 0 {
		return OnlineReqMultiError(errors)
	}
	return nil
}

// OnlineReqMultiError is an error wrapping multiple validation errors returned
// by OnlineReq.ValidateAll() if the designated constraints aren't met.
type OnlineReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OnlineReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OnlineReqMultiError) AllErrors() []error { return m }

// OnlineReqValidationError is the validation error returned by
// OnlineReq.Validate if the designated constraints aren't met.
type OnlineReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OnlineReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OnlineReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OnlineReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OnlineReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OnlineReqValidationError) ErrorName() string { return "OnlineReqValidationError" }

// Error satisfies the builtin error interface
func (e OnlineReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOnlineReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OnlineReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OnlineReqValidationError{}

// Validate checks the field values on OfflineReq with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OfflineReq) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OfflineReq with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OfflineReqMultiError, or
// nil if none found.
func (m *OfflineReq) ValidateAll() error {
	return m.validate(true)
}

func (m *OfflineReq) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.GetUid() <= 0 {
		err := OfflineReqValidationError{
			field:  "Uid",
			reason: "value must be greater than 0",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Key

	// no validation rules for Server

	if len(errors) > 0 {
		return OfflineReqMultiError(errors)
	}
	return nil
}

// OfflineReqMultiError is an error wrapping multiple validation errors
// returned by OfflineReq.ValidateAll() if the designated constraints aren't met.
type OfflineReqMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OfflineReqMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OfflineReqMultiError) AllErrors() []error { return m }

// OfflineReqValidationError is the validation error returned by
// OfflineReq.Validate if the designated constraints aren't met.
type OfflineReqValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OfflineReqValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OfflineReqValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OfflineReqValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OfflineReqValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OfflineReqValidationError) ErrorName() string { return "OfflineReqValidationError" }

// Error satisfies the builtin error interface
func (e OfflineReqValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOfflineReq.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OfflineReqValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OfflineReqValidationError{}

// Validate checks the field values on RegisterReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *RegisterReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RegisterReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in RegisterReplyMultiError, or
// nil if none found.
func (m *RegisterReply) ValidateAll() error {
	return m.validate(true)
}

func (m *RegisterReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return RegisterReplyMultiError(errors)
	}
	return nil
}

// RegisterReplyMultiError is an error wrapping multiple validation errors
// returned by RegisterReply.ValidateAll() if the designated constraints
// aren't met.
type RegisterReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RegisterReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RegisterReplyMultiError) AllErrors() []error { return m }

// RegisterReplyValidationError is the validation error returned by
// RegisterReply.Validate if the designated constraints aren't met.
type RegisterReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RegisterReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RegisterReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RegisterReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RegisterReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RegisterReplyValidationError) ErrorName() string { return "RegisterReplyValidationError" }

// Error satisfies the builtin error interface
func (e RegisterReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRegisterReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RegisterReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RegisterReplyValidationError{}

// Validate checks the field values on OnlineReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *OnlineReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on OnlineReply with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in OnlineReplyMultiError, or
// nil if none found.
func (m *OnlineReply) ValidateAll() error {
	return m.validate(true)
}

func (m *OnlineReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Uid

	// no validation rules for Key

	if len(errors) > 0 {
		return OnlineReplyMultiError(errors)
	}
	return nil
}

// OnlineReplyMultiError is an error wrapping multiple validation errors
// returned by OnlineReply.ValidateAll() if the designated constraints aren't met.
type OnlineReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m OnlineReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m OnlineReplyMultiError) AllErrors() []error { return m }

// OnlineReplyValidationError is the validation error returned by
// OnlineReply.Validate if the designated constraints aren't met.
type OnlineReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e OnlineReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e OnlineReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e OnlineReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e OnlineReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e OnlineReplyValidationError) ErrorName() string { return "OnlineReplyValidationError" }

// Error satisfies the builtin error interface
func (e OnlineReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sOnlineReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = OnlineReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = OnlineReplyValidationError{}

// Validate checks the field values on ServerIDReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ServerIDReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ServerIDReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ServerIDReplyMultiError, or
// nil if none found.
func (m *ServerIDReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ServerIDReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ServerID

	if len(errors) > 0 {
		return ServerIDReplyMultiError(errors)
	}
	return nil
}

// ServerIDReplyMultiError is an error wrapping multiple validation errors
// returned by ServerIDReply.ValidateAll() if the designated constraints
// aren't met.
type ServerIDReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ServerIDReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ServerIDReplyMultiError) AllErrors() []error { return m }

// ServerIDReplyValidationError is the validation error returned by
// ServerIDReply.Validate if the designated constraints aren't met.
type ServerIDReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ServerIDReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ServerIDReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ServerIDReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ServerIDReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ServerIDReplyValidationError) ErrorName() string { return "ServerIDReplyValidationError" }

// Error satisfies the builtin error interface
func (e ServerIDReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sServerIDReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ServerIDReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ServerIDReplyValidationError{}

// Validate checks the field values on ServerIDsReply with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ServerIDsReply) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ServerIDsReply with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ServerIDsReplyMultiError,
// or nil if none found.
func (m *ServerIDsReply) ValidateAll() error {
	return m.validate(true)
}

func (m *ServerIDsReply) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ServerIDsReplyMultiError(errors)
	}
	return nil
}

// ServerIDsReplyMultiError is an error wrapping multiple validation errors
// returned by ServerIDsReply.ValidateAll() if the designated constraints
// aren't met.
type ServerIDsReplyMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ServerIDsReplyMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ServerIDsReplyMultiError) AllErrors() []error { return m }

// ServerIDsReplyValidationError is the validation error returned by
// ServerIDsReply.Validate if the designated constraints aren't met.
type ServerIDsReplyValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ServerIDsReplyValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ServerIDsReplyValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ServerIDsReplyValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ServerIDsReplyValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ServerIDsReplyValidationError) ErrorName() string { return "ServerIDsReplyValidationError" }

// Error satisfies the builtin error interface
func (e ServerIDsReplyValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sServerIDsReply.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ServerIDsReplyValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ServerIDsReplyValidationError{}

// Validate checks the field values on UserToken with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UserToken) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UserToken with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in UserTokenMultiError, or nil
// if none found.
func (m *UserToken) ValidateAll() error {
	return m.validate(true)
}

func (m *UserToken) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UserTokenValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UserTokenValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UserTokenValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Token

	if len(errors) > 0 {
		return UserTokenMultiError(errors)
	}
	return nil
}

// UserTokenMultiError is an error wrapping multiple validation errors returned
// by UserToken.ValidateAll() if the designated constraints aren't met.
type UserTokenMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserTokenMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserTokenMultiError) AllErrors() []error { return m }

// UserTokenValidationError is the validation error returned by
// UserToken.Validate if the designated constraints aren't met.
type UserTokenValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserTokenValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserTokenValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserTokenValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserTokenValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserTokenValidationError) ErrorName() string { return "UserTokenValidationError" }

// Error satisfies the builtin error interface
func (e UserTokenValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUserToken.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserTokenValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserTokenValidationError{}

// Validate checks the field values on Userinfo with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *Userinfo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on Userinfo with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in UserinfoMultiError, or nil
// if none found.
func (m *Userinfo) ValidateAll() error {
	return m.validate(true)
}

func (m *Userinfo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Username

	// no validation rules for Nickname

	// no validation rules for Phone

	// no validation rules for Email

	// no validation rules for Sign

	// no validation rules for Avatar

	// no validation rules for Area

	// no validation rules for Gender

	if len(errors) > 0 {
		return UserinfoMultiError(errors)
	}
	return nil
}

// UserinfoMultiError is an error wrapping multiple validation errors returned
// by Userinfo.ValidateAll() if the designated constraints aren't met.
type UserinfoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserinfoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserinfoMultiError) AllErrors() []error { return m }

// UserinfoValidationError is the validation error returned by
// Userinfo.Validate if the designated constraints aren't met.
type UserinfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserinfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserinfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserinfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserinfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserinfoValidationError) ErrorName() string { return "UserinfoValidationError" }

// Error satisfies the builtin error interface
func (e UserinfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUserinfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserinfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserinfoValidationError{}
